File: Python/group-anagrams.py
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # Time:  O(n)
        # Space: O(n)
        bag = {}
        for s in strs:
            key = ''.join(sorted(s))
            bag.setdefault(key, list())
            bag[key].append(s)
        return list(bag.values())

File: C++/group-anagrams.cpp
#include <map>
#include <algorithm>
#include <string>
#include <vector>
#include <map>


class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(const std::vector<std::string>& strs) {
        // Time:  O(n)
        // Space: O(n)
        auto bag = std::map<std::string, std::vector<std::string>>();
        for (const std::string& s: strs) {
            std::string key = std::string(s);
            sort(key.begin(), key.end());
            bag[key].push_back(s);
        }
        std::vector<std::vector<std::string>> result = std::vector<std::vector<std::string>>();
        for (auto it: bag) {
            result.push_back(it.second);
        }
        return result;
    }
};